### PR TITLE
revise contract to suit frontend tutorial

### PR DESCRIPTION
**Pull request by pumpedlunch**
Insurance.sol was revised to better suit a frontend tutorial. Please consider updating the master version for consistency with the frontend tutorial.

Test files have not yet been updated for the Insurance.sol revisions

summary of changes:
1.  replaced issuePolicy with createPolicy, so that policies can be permissionlessly bought by a beneficiary rather than an insurer creating the policy. This allows for a user interactive frontend. This change required making the contract ownable, so an owner can set pricing for insurable events and withdraw funds deposited into the contract to cover policy liabilities
2. added a Status Enum and assertionId to the Policy struct. This simplified the frontend so that all relevant policy data can be obtained from the policies mapping rather than querying the InsurancePayoutRequested and InsurancePayoutSettled events and writing logic to sort through the events. The changes make requesting and settling payouts less gas efficient.

### CODE REVIEW

**Comment by pumpedlunch** [comment]
Point 1 above adds complexity to the Insurance.sol contract, requires more effort to revise, and is not critical to the frontend tutorial.
Point 2 is more important to the frontend tutorial, does not significantly increase the complexity of Insurance.sol and is low effort to change. 

As such, it may be best to omit bullet point 1, but make the changes for point 2. If agreed, I can revise this pull request as such.

File: src/Insurance.sol [hunk 1]
```diff
@@ -37,38 +47,45 @@ contract Insurance {
...
         defaultIdentifier = oo.defaultIdentifier();
     }
 
-    function issueInsurance(
+    function createPolicy(
         uint256 insuranceAmount,
         address payoutAddress,
         bytes memory insuredEvent
     ) public returns (bytes32 policyId) {
+        require(insuredEventPrice[insuredEvent] > 0, "insuredEvent invalid");
```
Comment (Reinis-FRP) [inline src/Insurance.sol:63]: insuredEvent is bytes, not bytes32 as expected in the mapping

File: src/Insurance.sol [hunk 2]
```diff
@@ -4,27 +4,37 @@ pragma solidity ^0.8.0;
 import "@uma/core/contracts/optimistic-oracle-v3/implementation/ClaimData.sol";
 import "@uma/core/contracts/optimistic-oracle-v3/interfaces/OptimisticOracleV3Interface.sol";
 import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
 
-// This Isurance contract enables for the issuance of a single unlimited time policy per event/payout recipient There is
-// no limit to the number of payout requests that can be made of the same policy; however, only the first asserted
-// request will settle the insurance payment, whereas OOv3 will settle bonds for all requestors.
-contract Insurance {
+contract InsuranceTutorial is Ownable{
```
Comment (Reinis-FRP) [inline src/Insurance.sol:9]: Renaming breaks tests

File: src/Insurance.sol [hunk 3]
```diff
@@ -37,38 +47,45 @@ contract Insurance {
...
+    event InsuredEventSet(bytes32 indexed insuredEvent, uint256 price);
+
     constructor(address _defaultCurrency, address _optimisticOracleV3) {
         defaultCurrency = IERC20(_defaultCurrency);
         oo = OptimisticOracleV3Interface(_optimisticOracleV3);
         defaultIdentifier = oo.defaultIdentifier();
     }
 
-    function issueInsurance(
+    function createPolicy(
```
Comment (Reinis-FRP) [inline src/Insurance.sol:58]: renaming breaks tests

File: src/Insurance.sol [hunk 4]
```diff
@@ -37,38 +47,45 @@ contract Insurance {
...
         policies[policyId] = Policy({
             insuranceAmount: insuranceAmount,
             payoutAddress: payoutAddress,
             insuredEvent: insuredEvent,
-            settled: false
+            status: Status.OPEN,
+            assertionId : ''
         });
-        defaultCurrency.safeTransferFrom(msg.sender, address(this), insuranceAmount);
+        defaultCurrency.safeTransferFrom(msg.sender, address(this), insuranceAmount * insuredEventPrice[insuredEvent] / 10000 );
```
Comment (Reinis-FRP) [inline src/Insurance.sol:74]: insuredEvent is bytes, not bytes32 as expected in the mapping

File: src/Insurance.sol [hunk 5]
```diff
@@ -83,14 +100,34 @@ contract Insurance {
             bytes32(0) // No domain.
         );
         assertedPolicies[assertionId] = policyId;
+
+        policy.status = Status.REQUESTED;
+        policy.assertionId = assertionId;
         emit InsurancePayoutRequested(policyId, assertionId);
     }
+    // note: price in BPS, insuredEvents with price = 0 are invalid
```
Comment (Reinis-FRP) [inline src/Insurance.sol:108]: you might want add require that price > 0

File: src/Insurance.sol [hunk 6]
```diff
@@ -4,27 +4,37 @@ pragma solidity ^0.8.0;
...
+        bytes32 assertionId;
     }
 
     mapping(bytes32 => bytes32) public assertedPolicies;
 
     mapping(bytes32 => Policy) public policies;
+    // insuredEvent => BPS price per policy amount
+    mapping(bytes32 => uint256) public insuredEventPrice;
+
+    uint256 public totalLiabilities;
```
Comment (Reinis-FRP) [inline src/Insurance.sol:37]: this is never assigned, so its 0 in all checks below

File: src/Insurance.sol (hunk 4 above)
Comment (Reinis-FRP) [inline src/Insurance.sol:74]: Is insuredEventPrice[insuredEvent] / 10000 < 1? If so, the settlement would revert as it requires full insuranceAmount payout

**Review by Reinis-FRP — COMMENTED** [review]
Happy to see people building on these examples!

Though, I would suggest closing this PR as we would like to keep the code as much bare-bones as possible in this repo as it solely serves the purpose of demonstrating OOv3 integration points and is not meant to be a complete product.

Anyways, I left couple of comments that can be considered in your fork and also would suggest elaborating in tests to support your changes.

**Comment by pumpedlunch** [comment]
Thanks Reinis, I will close this draft PR, incorporate your comments to the revised contract and share the revised contract in my frontend tutorial repo.

**Pull request closed without merging**